Add fabric gateway and routing protocol terraform provider

manu-equinix opened the pull request:
- Adds new Fabrc *resources* and *datasources*:
  - `equinix_fabric_gateway`
  - `equinix_fabric_routingprotocol`
- Replaces Datasource `equinix_fabric_connection` `type` attribute value `GW_VC` with `IP_VC`
- Includes terraform examples for the above 2 providers and a shell script to test the end to end workflow.
- Updates Fabric-go client SDK to version v0.5.0
- PORT : A SET(nested schema) has been removed and simplified into a boolean `LagEnabled` on the resource `equinix_fabric_port` - More information about the API can be found at https://developer.equinix.com/catalog/fabricv4#operation/createPort
- Collaborators: @manu-equinix, @sabwong, @xbai-equinix, @kehuang114

ctreatma commented on equinix/fabric_port_mapping_helper.go:
```diff
@@ -104,26 +104,26 @@ func encapsulationToTerra(portEncapsulation *v4.PortEncapsulation) *schema.Set {
-	)
-	return portLagSet
-}
+//func lagToTerra(portLag *v4.PortLag) *schema.Set {
```
Deleting code is preferable to commenting it out; if we needed to bring it back later we can always find it in the git history.

manu-equinix replied:
Ack!

ctreatma reviewed (commented):
It looks like there are a couple new data sources and resources, but I don't see new docs for those.  Could you add them?

Re: tests, how can we run the fabric tests to validate that new & existing tests still pass?

manu-equinix commented:
> It looks like there are a couple new data sources and resources, but I don't see new docs for those. Could you add them?
> 
> Re: tests, how can we run the fabric tests to validate that new & existing tests still pass?

Added docs for the new data sources and resources.  

Re: tests - We have added tests for the new resources under "tests" folder? 
Additionally, we also have a few example .tf scripts which we ran to verify the functionality of the newly added Fabric gateway and the routing protocol providers. Also, added a `examples/fabric-gateway/test.sh` which contains end to end workflow for resource creation to deletion.

Re: the existing resources, I believe we have tests under the tests folder.

displague commented on examples/fabric-gateway/test.sh:
```diff
@@ -0,0 +1,89 @@
+#--------create FG
```
A file named `test.sh` is unconventional. This may be better as a README.md or a TESTING.md that breaks this document up into explanations and headings that perform the steps outlined below.

What we really need is an E2E testing credential for Fabric resources in the Terraform provider.

manu-equinix replied:
Changed this to a markdown file and formatted accordingly.

displague replied:
The formatting could be improved.

* the code blocks could be captured in blocks like this:
   ~~~
   ```sh
   terraform something
   ```
   ~~~
* the filename could be TEST.md
* the headings could be in better descending order: `#` (only at the top), `##` (for sub sections), `###` (for sub-sub sections)
* newlines could be added between the headings, sub-headings, and code-blocks

We should add markdown linters for these concerns though, I won't fault you for it. We can chase these improvements in future PRs.

manu-equinix replied:
Ack. Thanks for pointing out. This was supposed to be an executable shell script for internal testing. I missed some formatting due to last minute changes.  I have now updated this file with the correct formatting.

displague commented on docs/data-sources/equinix_fabric_routingprotocol.md:
```diff
@@ -0,0 +1,165 @@
+
+Fabric V4 API compatible data resource that allow user to fetch routing protocol for a given UUID
+
+~> **Note** Equinix Fabric v4 resources and datasources are currently in Beta. The interfaces related to `equinix_fabric_` resources and datasources may change ahead of general availability. Please, do not hesitate to report any problems that you experience by opening a new [issue](https://github.com/equinix/terraform-provider-equinix/issues/new?template=bug.md)
```
Is this still the case? When can we call the Equinix Fabric API, resources, and datasources stable?

manu-equinix replied:
The Fabric Gateway(which is being added here) product is still under development and a beta release will be available soon. We can mark it stable in the next few months or whenever we go GA with this product.

displague commented on docs/data-sources/equinix_fabric_routingprotocol.md:
```diff
@@ -0,0 +1,165 @@
+
+# equinix_fabric_routingprotocol (Data Source)
+
+Fabric V4 API compatible data resource that allow user to fetch routing protocol for a given UUID
```
Can we link this to the docs for Fabric V4 Routing Protocol? (preferably both the product docs and the API docs, but at least the latter)

displague commented on docs/resources/equinix_fabric_gateway.md:
```diff
@@ -0,0 +1,137 @@
+
+# equinix_fabric_gateway (Resource)
+
+Fabric V4 API compatible resource allows creation and management of Equinix Fabric Gateway
```
Same comment here about linking to the product docs or the API docs within this brief explanation.

displague commented on docs/resources/equinix_fabric_gateway.md:
```diff
@@ -0,0 +1,137 @@
+
+Fabric V4 API compatible resource allows creation and management of Equinix Fabric Gateway
+
+~> **Note** Equinix Fabric v4 resources and datasources are currently in Beta. The interfaces related to `equinix_fabric_` resources and datasources may change ahead of general availability. Please, do not hesitate to report any problems that you experience by opening a new [issue](https://github.com/equinix/terraform-provider-equinix/issues/new?template=bug.md)
```
Same comment here about wether and for how long this disclaimer applies.

displague commented on equinix/fabric_port_read_schema.go:
```diff
@@ -308,12 +308,9 @@ func readFabricPortResourceSchema() map[string]*schema.Schema {
 		},
 		"lag": {
-			Type:        schema.TypeSet,
+			Type:        schema.TypeBool,
```
This change should be included in the docs and included in the PR description

displague replied:
Looks like it affects the datasource for fabric_port and fabric_ports.

https://registry.terraform.io/providers/equinix/equinix/latest/docs/data-sources/equinix_fabric_port#nested-schema-for-lag

manu-equinix replied:
Added link to API documentation and more information about this change to the PR description.

ctreatma replied:
I think it would be preferable to add a new `lag_enabled` property, which lines up with the new `lagEnabled` property in the API spec.  Has support for the old `lag` property been dropped entirely?  If a customer has an existing terraform configuration that uses
```hcl
lag {
  enabled = true
  name = "some name"
  # ...
}
```
and they run `terraform plan` / `terraform apply` right now, using the current release of the terraform provider, will they get an error or does is still work?

manu-equinix replied:
Yes, the API spec has been updated from the older implementation and the customers might need to change their terraform scripts to accommodate the changed fields. 
cc @anchala71

anchala71 replied:
There is no Customer for the current version of Port specifications and we did release for a beta versions. I think it would be safe to deprecate the field. Anyways let me add team who involved in this change @askamanth-equinix

anchala71 replied:
@askamanth-equinix

manu-equinix replied:
@ctreatma Ack. I have updated the property to lag_enabled.

prashanthatequinix replied:
Since no customer is using this feature as of now, we can go ahead and release this.

manu-equinix replied:
@ctreatma I have updated this field now.

ctreatma commented on examples/fabric-gateway/TEST.md:
````diff
@@ -0,0 +1,155 @@
+
+## Create FG
+
+```cd fabric-gateway/
````
Are there manual steps that happen outside of terraform for these tests (for example, a person has to respond to a ticket or click a button in order to complete the provisioning process for a resource)?  If it's all terraform, it would be preferable to create a test case or cases like the others in the `tests/` directory.

manu-equinix replied:
This was just a helper shell script(Now modified into a .md file) we created for the demo in addition to actual terraform tests under `/tests` directory.

ctreatma replied:
I see that this markdown file was created in response to an earlier comment on the shell script.  Given the current state of end-to-end testing for these resources, I think we can merge this doc as-is.

ctreatma commented:
This needs to be rebased on the latest from `main` to resolve conflicts.

manu-equinix commented:
@ctreatma Just did the rebase on master and fixed all merge conflicts.
